src/jvm/main/org/jetbrains/kotlinx/lincheck/paramgen/ExpandingRangeIntGenerator.kt

```diff
@@ -0,0 +1,163 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2023 JetBrains s.r.o.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>
+ */
+
+package org.jetbrains.kotlinx.lincheck.paramgen
+
+import java.util.*
+
+
+/**
+ * Extensible range-based random generator.
+ *
+ * @param random   random for this generator
+ * @param startInclusive    initial range left bound (inclusive)
+ * @param endInclusive      initial range right bound (inclusive)
+ * @param minStartInclusive minimal range left bound (inclusive)
+ * @param maxEndInclusive   maximal range right bound (inclusive)
+ */
+class ExpandingRangeIntGenerator(
+    private val random: Random,
+    private val startInclusive: Int,
```

Comment:
What is the difference between startInclusive and minStartInclusive? Why do you need both?

Reply:
`startInclusive` stands for initial range leftBound, when `minStartInclusive` is the lowest bound of range expansion.
For example if `startInclusive` = 3, and `minStartInclusive` = 1,  `endInclusive` = 5, and `maxEndInclusive` = 7, range will initially be [3, 5], and eventually will expand to [1, 7].
I have to store them both as we need to know if we can expand range or not.